utils/disk_space_monitor.hh

```diff
@@ -0,0 +1,127 @@
+
+#include <chrono>
+#include <filesystem>
+#include <set>
+
+#include <boost/intrusive/list.hpp>
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include "seastarx.hh"
+
+using namespace std::chrono_literals;
+
+namespace bi = boost::intrusive;
+
+namespace utils {
+
+// Instantiated only on shard 0
+class disk_space_monitor {
+public:
+    struct config {
+        scheduling_group sched_group;
+        lowres_clock::duration normal_polling_interval = 10s;
+        lowres_clock::duration high_polling_interval = 1s;
+        // Use high_polling_interval above this threshold
+        float polling_interval_threshold = 0.95;
+    };
+
+    using subscription_callback_type = noncopyable_function<future<> (const disk_space_monitor&)>;
+    class subscription : public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
+        friend class disk_space_monitor;
+
+        subscription_callback_type _callback;
+
+        explicit subscription(subscription_callback_type callback)
+                : _callback(std::move(callback)) {
+        }
+
+    public:
+        subscription() = default;
+
+        subscription(subscription&& other) noexcept
+                : _callback(std::move(other._callback))
+        {
+            subscription_list_type::node_algorithms::swap_nodes(other.this_ptr(), this_ptr());
+        }
+
+        subscription& operator=(subscription&& other) noexcept {
+            if (this != &other) {
+                _callback = std::move(other._callback);
+                unlink();
+                subscription_list_type::node_algorithms::swap_nodes(other.this_ptr(), this_ptr());
+            }
+            return *this;
+        }
+
+        explicit operator bool() const noexcept {
+            return is_linked();
+        }
+
+        future<> on_event(const disk_space_monitor& dsm) const noexcept {
+            return _callback(dsm);
+        }
+    };
+
+private:
+    std::set<std::filesystem::path> _data_dirs;
+    config _cfg;
+    timer<lowres_clock> _timer;
+    uint64_t _total_space = 0;
+    uint64_t _free_space = 0;
+    gate _async_gate;
+
+    using subscription_list_type = bi::list<subscription, bi::constant_time_size<false>>;
+    subscription_list_type _subscriptions;
+
+public:
+    void start(std::set<std::filesystem::path> data_dirs, config cfg);
+    void start(std::filesystem::path data_dir, config cfg) {
+        start(std::set<std::filesystem::path>({std::move(data_dir)}), std::move(cfg));
+    }
+    future<> stop() noexcept;
+
+    const std::set<std::filesystem::path>& data_dirs() const noexcept {
+        return _data_dirs;
+    }
+    uint64_t total_space() const noexcept { return _total_space; }
+    uint64_t free_space() const noexcept { return _free_space; }
+    float disk_utilization() const noexcept {
+        return _total_space ? (float)(_total_space - _free_space) / _total_space : -1;
```

Comment:
This isn't meaningful, each mount has to be managed separately.

Your average utilization can be 80% and one of the mounts at 100%.

Reply:
okay, though it's all theoretical.
We currently support only a single dir in pratice.
Maybe for simplicity reasons we can start with a single dir, and extend it to multiple directories if and when we really support that end-to-end.

Reply:
Let's not add more breakage.

If this thing supports multiple directories, it should support them correctly.

Reply:
We don't have to support multiple directories at the moment.

Reply:
another possible usage of disk monitor is object storage. we monitor fast storage, and if filling up, we throttle incoming writes in order to allow data to be demoted to slow storage without ENOSPC. /cc @tchaikov